### PR TITLE
Add RLHF 

### PR DESCRIPTION
**Pull request by hxdtest**
### What changes were proposed in this pull request?
We add ATorch RLHF for large language model fine tuning. This pr fix some problems for deepspeed hybrid engine, enable training model from 7B to 70B using zero2( zero2 training and single gpu inferencing) and hybrid engine(zero3 training + tensor parallel inferencing).

As for RLHF, training data needs to be generated using latest model weights. If  we could increase data generation speed, total training time would be reduced. In addition, there are so many inferencing backend, like vllm/tgi/onnx_runtime.  ATorch RLHF  is designed support various inferencing backend by gathering trainable model weights and sending them to inference instance without creating torch process group between inference instance and training process.

### Why are the changes needed?

Deepspeed hybrid engine is an solution for large language model training and inferencing but encounters some bugs when used for RLHF training [issues821](https://github.com/microsoft/DeepSpeedExamples/issues/821)  [issues643](https://github.com/microsoft/DeepSpeedExamples/issues/643) [issues628](https://github.com/microsoft/DeepSpeedExamples/issues/628). We fix tp weight slicing error, applying tensor parallelism error and replace deepspeed custome op with native torch op.

### Does this PR introduce any user-facing change?

Specify whether this pull request introduces any changes that users will directly interact with or notice.

### How was this patch tested?

Detail the testing process you have undertaken to ensure the changes in this pull request are valid and working as intended.

### CODE REVIEW

**Comment by workingloong** [comment]
We should name the PR with a specific title.

File: atorch/atorch/__init__.py [hunk 1]
```diff
@@ -1,14 +1,8 @@
 import logging
 import os
-from importlib.metadata import version
```
Comment (skydoorkai) [inline atorch/atorch/__init__.py:3]: Don't change this file.

File: atorch/atorch/auto/accelerate.py [hunk 2]
```diff
@@ -12,6 +12,8 @@
 from atorch.auto.auto_accelerate_context import AutoAccelerateContext
 from atorch.auto.device_context import get_device_context
 from atorch.auto.dry_runner.dry_runner import get_dryrunner
+from atorch.auto.engine.acceleration_engine import AccelerationEngine
```
Comment (skydoorkai) [inline atorch/atorch/auto/accelerate.py:15]: Also, merge to the latest version, not revert back to old version.

File: atorch/atorch/auto/model_context.py [hunk 3]
```diff
@@ -383,8 +383,8 @@ def create_optim(self):
 
         # model must on cuda device before calling OSS (zero2)
         if "zero2" in self.post_wrappers or "ds_zero" in self.post_wrappers:
-            is_cuda = next(self.model.parameters()).is_cuda
-            if torch.cuda.is_available() and not is_cuda:
+            model_device = next(self.model.parameters()).device
```
Comment (skydoorkai) [inline atorch/atorch/auto/model_context.py:386]: Same this file.

**Review by skydoorkai — COMMENTED** [review]
Merge RLHF codes to latest atorch version.

File: atorch/atorch/auto/model_context.py [hunk 4]
```diff
@@ -430,9 +430,9 @@ def create_optim(self):
             and "fsdp" not in self.pre_wrappers
             and "ds_3d_parallel" not in self.post_wrappers
         ):
-            is_cuda = next(self.model.parameters()).is_cuda
+            model_device = next(self.model.parameters()).device
             # model must on cuda device before calling BF16Optimizer
-            if torch.cuda.is_available() and not is_cuda:
+            if torch.cuda.is_available() and model_device.type != "cuda":
```
Comment (linkerzhang) [inline atorch/atorch/auto/model_context.py:435]: this check appears twice so far, refactor the codes to do it once only.

**Comment by skydoorkai** [comment]
close as corresponding codes have already be merged to master.

**Pull request closed without merging**